framework/src/problems/FEProblemBase.C

```diff
@@ -7786,11 +7786,13 @@ FEProblemBase::checkNonlinearConvergence(std::string & msg,
 
   if ((it >= _nl_forced_its) && it && reason == MooseNonlinearConvergenceReason::ITERATING)
   {
-    // If compute_initial_residual_before_preset_bcs==false, then use the
-    // first residual computed by PETSc to determine convergence.
-    Real the_residual = system._compute_initial_residual_before_preset_bcs
-                            ? initial_residual_before_preset_bcs
-                            : system._initial_residual_after_preset_bcs;
+    // Set the initial residual depending on what the user asks us to use.
+    Real the_residual;
+    if (_app.parameters().get<bool>("use_legacy_initial_residual_evaluation_bahavior"))
+      the_residual = system._use_fnorm0_before_smo ? fnorm0_before_smo : system._fnorm0_after_smo;
+    else
+      the_residual =
+          system.shouldEvaluateFNormBeforeSMO() ? fnorm0_before_smo : system._fnorm0_after_smo;
```

Comment:
I am not sure if it is possible. The main issue is that the legacy behavior is not self consistent: it chooses which residual to use according to this input parameter, but it always performs the redundant residual evaluation (even if it will not be used).